### PR TITLE
Dynamic test

### PR DESCRIPTION
**Pull request by tballmsft**

### CODE REVIEW

File: src/testdom/testrules.ts [hunk 1]
```diff
@@ -169,7 +173,53 @@ export function resolveTestRules(serviceClass: number) {
     return builtinTestRules[serviceClass]
 }
 
+function makeEventTest(test: ServiceMemberTestNode, event:string, flag: number) {
```
Comment (pelikhan) [inline src/testdom/testrules.ts:176]: The others are called "create"

File: src/testdom/nodes.ts [hunk 2]
```diff
@@ -586,6 +586,7 @@ export class RegisterTest extends RegisterTestNode {
 }
 
 export class EventTest extends ServiceMemberTestNode {
+    public seen = false
```
Comment (pelikhan) [inline src/testdom/nodes.ts:589]: Probably can be part of the closure down the road.

File: src/testdom/testrules.ts [hunk 3]
```diff
@@ -169,7 +173,53 @@ export function resolveTestRules(serviceClass: number) {
...
+            }
+        )
+    )
+}
+
 const builtinServiceCommandTests: Record<number, ServiceMemberOptions> = {
+    [SRV_GAMEPAD]: {
+        name: "gamepad events",
+        start: test => {
+            const service = test.service
```
Comment (pelikhan) [inline src/testdom/testrules.ts:196]: I think instead you should register to the REPORT_UPDATE event and refresh when you get a value, return the unsubscribe function.

File: src/testdom/testrules.ts [hunk 4]
```diff
@@ -169,7 +173,53 @@ export function resolveTestRules(serviceClass: number) {
...
+                    if (!isNaN(value)) {
+                        if (value & buttons) {
+                            makeEventTest(test, key, value)
+                        }
+                    }
+                }
+                // if buttons doesn't have any of L/R/D/U, then add 
+                // as we have a analog joystick 
+                const LRUD: number[] = [ GamepadButtons.Down , GamepadButtons.Up, GamepadButtons.Left, GamepadButtons.Right ]
+                if (!(buttons & LRUD.reduceRight((prev,curr) => (prev | curr),0))) {
```
Comment (pelikhan) [inline src/testdom/testrules.ts:211]: Use bitmask &

File: src/testdom/testrules.ts [hunk 5]
```diff
@@ -169,7 +173,53 @@ export function resolveTestRules(serviceClass: number) {
     return builtinTestRules[serviceClass]
 }
 
+function makeEventTest(test: ServiceMemberTestNode, event:string, flag: number) {
+    test.appendChild(
+        new EventTest(
+            `${event}`,
+            GamepadEvent.ButtonsChanged,
+            (node, logger) => {
+                const { event } = node
```
Comment (pelikhan) [inline src/testdom/testrules.ts:182]: Isn't there an event test node that does this already?

**Review by pelikhan — CHANGES_REQUESTED** [review]
Some comments.

File: src/testdom/testrules.ts (hunk 3 above)
Comment (tballmsft) [reply src/testdom/testrules.ts:196]: hmm... not clear that this is going to work - the register has only one value, fixed for the hardware.

File: src/testdom/testrules.ts (hunk 5 above)
Comment (tballmsft) [reply src/testdom/testrules.ts:182]: yes, using it already.

File: src/testdom/testrules.ts (hunk 3 above)
Comment (tballmsft) [reply src/testdom/testrules.ts:215]: ok - i get it - working on it.

**Pull request merged**